server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java

```diff
@@ -0,0 +1,278 @@
+    ) throws IOException {
+        super(name, factories, context, parent, CardinalityUpperBound.MANY, metadata);
+        this.config = config;
+        this.keyed = keyed;
+        this.minDocCount = minDocCount;
+        this.bucketOrds = BytesKeyedBucketOrds.build(bigArrays(), cardinality);
+        this.ipPrefix = ipPrefix;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return config.getValuesSource() == null
+            ? LeafBucketCollector.NO_OP_COLLECTOR
+            : new IpPrefixLeafCollector(sub, config.getValuesSource().bytesValues(ctx), ipPrefix);
+    }
+
+    private class IpPrefixLeafCollector extends LeafBucketCollectorBase {
+        private final IpPrefix ipPrefix;
+        private final LeafBucketCollector sub;
+        private final SortedBinaryDocValues values;
+
+        IpPrefixLeafCollector(LeafBucketCollector sub, SortedBinaryDocValues values, IpPrefix ipPrefix) {
+            super(sub, values);
+            this.sub = sub;
+            this.values = values;
+            this.ipPrefix = ipPrefix;
+        }
+
+        @Override
+        public void collect(int doc, long owningBucketOrd) throws IOException {
+            BytesRef previousSubnet = null;
+            BytesRef subnet;
+            BytesRef ipAddress;
+            if (values.advanceExact(doc)) {
+                int valuesCount = values.docValueCount();
+
+                for (int i = 0; i < valuesCount; ++i) {
+                    ipAddress = values.nextValue();
+                    subnet = maskIpAddress(ipAddress, ipPrefix.netmask);
+                    if (previousSubnet != null && subnet.bytesEquals(previousSubnet)) {
+                        continue;
+                    }
+                    long bucketOrd = bucketOrds.add(owningBucketOrd, subnet);
+                    if (bucketOrd < 0) {
+                        bucketOrd = -1 - bucketOrd;
+                        collectExistingBucket(sub, doc, bucketOrd);
+                    } else {
+                        collectBucket(sub, doc, bucketOrd);
+                    }
+                    previousSubnet = subnet;
+                }
+            }
+        }
+
+        private BytesRef maskIpAddress(final BytesRef ipAddress, final BytesRef netmask) {
+            assert ipAddress.length == 16 : "Invalid length for ip address [" + ipAddress.length + "] expected 16 bytes";
+            return mask(ipAddress, netmask);
+        }
+
+        private BytesRef mask(final BytesRef ipAddress, final BytesRef subnetMask) {
+            // NOTE: IPv4 addresses are encoded as 16-bytes. As a result, we use an
+            // offset (12) to apply the subnet to the last 4 bytes (byes 12, 13, 14, 15)
+            // if the subnet mask is just a 4-bytes subnet mask.
+            int offset = subnetMask.length == 4 ? 12 : 0;
+            byte[] subnet = new byte[subnetMask.length];
```

Comment:
I still don't think we should be allocating short lived objects here.  I think you can just pass in the `BytesRef` and mutate it in place.